src/hotspot/cpu/aarch64/aarch64_vector_ad.m4

```diff
@@ -0,0 +1,4693 @@
+                                    int opcode, Register base, int index, int size, int disp) {
+    sve_mem_insn_predicate insn;
+    int mesize = type2aelembytes(mem_elem_bt);
+    if (index == -1) {
+      assert(size == 0, "unsupported address mode: scale size = %d", size);
+      switch(mesize) {
+      case 1:
+        insn = is_store ? &C2_MacroAssembler::sve_st1b : &C2_MacroAssembler::sve_ld1b;
+        break;
+      case 2:
+        insn = is_store ? &C2_MacroAssembler::sve_st1h : &C2_MacroAssembler::sve_ld1h;
+        break;
+      case 4:
+        insn = is_store ? &C2_MacroAssembler::sve_st1w : &C2_MacroAssembler::sve_ld1w;
+        break;
+      case 8:
+        insn = is_store ? &C2_MacroAssembler::sve_st1d : &C2_MacroAssembler::sve_ld1d;
+        break;
+      default:
+        assert(false, "unsupported");
+        ShouldNotReachHere();
+      }
+      int imm4 = disp / mesize / Matcher::scalable_vector_reg_size(vector_elem_bt);
+      (masm.*insn)(reg, Assembler::elemType_to_regVariant(vector_elem_bt), pg, Address(base, imm4));
+    } else {
+      assert(false, "unimplemented");
+      ShouldNotReachHere();
+    }
+  }
+
+  const bool Matcher::match_rule_supported_superword(int opcode, int vlen, BasicType bt) {
+    if (UseSVE == 0) {
+      // ConvD2I and ConvL2F are not profitable to be vectorized on NEON, because no direct
+      // NEON instructions support them. But the match rule support for them is profitable for
+      // Vector API intrinsics.
+      if ((opcode == Op_VectorCastD2X && bt == T_INT) ||
+          (opcode == Op_VectorCastL2X && bt == T_FLOAT)) {
+        return false;
+      }
+    }
+    return match_rule_supported_vector(opcode, vlen, bt);
+  }
+
+  // Identify extra cases that we might want to provide match rules for vector nodes and
+  // other intrinsics guarded with vector length (vlen) and element type (bt).
+  const bool Matcher::match_rule_supported_vector(int opcode, int vlen, BasicType bt) {
+    if (!match_rule_supported(opcode)) {
+      return false;
+    }
+
+    int length_in_bytes = vlen * type2aelembytes(bt);
+    if (UseSVE == 0 && length_in_bytes > 16) {
+      return false;
+    }
+
+    switch (opcode) {
+      case Op_MulVL:
+      case Op_AndVMask:
+      case Op_OrVMask:
+      case Op_XorVMask:
+      case Op_MaskAll:
+      case Op_VectorMaskGen:
+      case Op_LoadVectorMasked:
+      case Op_StoreVectorMasked:
+      case Op_LoadVectorGather:
+      case Op_StoreVectorScatter:
+      case Op_LoadVectorGatherMasked:
+      case Op_StoreVectorScatterMasked:
+      case Op_PopulateIndex:
+      case Op_CompressM:
+      case Op_CompressV:
+        if (UseSVE == 0) {
+          return false;
+        }
+        break;
+      case Op_MulAddVS2VI:
+        return length_in_bytes == 16;
+      case Op_MulReductionVD:
+      case Op_MulReductionVF:
+      case Op_MulReductionVI:
+      case Op_MulReductionVL:
+        // No multiply reduction instructions, and we emit scalar
```

Comment:
This comment is a little unclear. Is this what you actually mean?

"No vector multiply reduction instructions, but we do emit scalar instructions for 64/128-bit vectors"

Reply:
Yes. You're right. Updated in the new revision. Thanks.